### PR TITLE
help: Generate checkpoint height from chainparams

### PR DESCRIPTION
**Pull request by luke-jr**
Not sure if this is worth putting in Core, but might as well until checkpoints are removed entirely.

### CODE REVIEW

**Comment by laanwj** [comment]
ACK 6ba3f6d49e09024af9bce7a184aaaf6caf766c40
This is preferable to hardcoding the value in the string.

File: src/init.cpp [hunk 1]
```diff
@@ -482,7 +482,7 @@ void SetupServerArgs()
...
     gArgs.AddArg("-checkmempool=<n>", strprintf("Run checks every <n> transactions (default: %u, regtest: %u)", defaultChainParams->DefaultConsistencyChecks(), regtestChainParams->DefaultConsistencyChecks()), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
-    gArgs.AddArg("-checkpoints", strprintf("Enable rejection of any forks from the known historical chain until block 295000 (default: %u)", DEFAULT_CHECKPOINTS_ENABLED), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    gArgs.AddArg("-checkpoints", strprintf("Enable rejection of any forks from the known historical chain until block %s (default: %s)", defaultChainParams->Checkpoints().mapCheckpoints.rbegin()->first, DEFAULT_CHECKPOINTS_ENABLED), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
Comment (donaloconnor) [inline src/init.cpp:485]: I'd prefer adding a function to CCheckpointData to obtain this rather than exposing the map details here.

Comment (laanwj) [reply src/init.cpp:485]: Good idea.

**Comment by luke-jr** [comment]
Abstracted checkpoint access and rebased

**Comment by sipa** [comment]
utACK 40ef5d4100c575db35fa736857d41a7c0c7df6de

**Comment by donaloconnor** [comment]
utACK https://github.com/bitcoin/bitcoin/commit/40ef5d4100c575db35fa736857d41a7c0c7df6de

File: src/init.cpp [hunk 2]
```diff
@@ -506,7 +506,7 @@ void SetupServerArgs(NodeContext& node)
...
     argsman.AddArg("-checkmempool=<n>", strprintf("Run checks every <n> transactions (default: %u, regtest: %u)", defaultChainParams->DefaultConsistencyChecks(), regtestChainParams->DefaultConsistencyChecks()), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
-    argsman.AddArg("-checkpoints", strprintf("Enable rejection of any forks from the known historical chain until block 295000 (default: %u)", DEFAULT_CHECKPOINTS_ENABLED), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-checkpoints", strprintf("Enable rejection of any forks from the known historical chain until block %s (default: %s)", defaultChainParams->Checkpoints().GetHeight(), DEFAULT_CHECKPOINTS_ENABLED), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
Comment (MarcoFalke) [inline src/init.cpp:509]: boolean values represented as strings are not supported by the argparser. We shouldn't imply that nor should this method be changed to be inconsistent with all other boolean default values

**Review by maflcko — CHANGES_REQUESTED** [review]

File: src/init.cpp (hunk 2 above)
Comment (luke-jr) [reply src/init.cpp:509]: Fixed

File: src/chainparams.h [hunk 3]
```diff
@@ -23,6 +23,11 @@ typedef std::map<int, uint256> MapCheckpoints;
 
 struct CCheckpointData {
     MapCheckpoints mapCheckpoints;
+
+    int GetHeight() const {
+        const auto& final_checkpoint = mapCheckpoints.rbegin();
+        return final_checkpoint->first /* height */;
```
Comment (jnewbery) [inline src/chainparams.h:29]: I think ideally, any begin/end/rbegin/rend iterator should be tested before deferencing:

```c++
return final_checkpoint == mapCheckpoints.rend() ? 0 : final_checkpoint->first;
```

or similar.

**Pull request merged**